hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -845,7 +846,52 @@ case class TableMapRows(child: TableIR, newRow: IR) extends TableIR {
         scanInitOps(0)(region, scanAggs, globals, false)
       }
 
-      val itF = { (i: Int, ctx: RVDContext, partitionAggs: Array[RegionValueAggregator], it: Iterator[RegionValue]) =>
+      val scanAggsPerPartition = SpillingCollectIterator(tv.rvd.mapPartitionsWithIndex { (i, ctx, it) =>
+        val globals =
+          if (scanSeqNeedsGlobals) {
+            val rvb = new RegionValueBuilder(ctx.freshRegion)
+            rvb.start(gType.physicalType)
+            rvb.addAnnotation(gType, globalsBc.value)
+            rvb.end()
+          } else
+              0
+
+        val scanSeqOpF = scanSeqOps(i)
+        it.foreach { rv =>
+          scanSeqOpF(rv.region, scanAggs, globals, false, rv.offset, false)
+          ctx.region.clear()
+        }
+        Iterator.single(scanAggs)
+      }, HailContext.get.flags.get("max_leader_scans").toInt).scanLeft(scanAggs) { (a1, a2) =>
+        (a1, a2).zipped.map { (agg1, agg2) =>
+          val newAgg = agg1.copy()
+          newAgg.combOp(agg2)
+          newAgg
+        }
+      }
+
+      val partitionIndices = new Array[Long](tv.rvd.getNumPartitions + 1)
+      var i = 0
+      val scanAggsPerPartitionFile = hc.getTemporaryFile()
+      HailContext.get.sFS.writeFileNoCompression(scanAggsPerPartitionFile) { os =>
+        scanAggsPerPartition.foreach { x =>
+          partitionIndices(i) = os.getPos
+          i += 1
+          val oos = new ObjectOutputStream(os)
```

Comment:
can we lift the oos up outside the loop, and flush once at the end?

Reply:
oh, you need to flush inside to make the position correct. Could you add a comment to that effect?

Reply:
yeah it was pretty weird, I should verify but I also had problems getting the right positions when re-using the same OOS